### DISCUSSION TITLE
ABRP Integration

### DISCUSSION THREAD

**Discussion by dthiery** [teslamate-org/teslamate#1592] [Ideas]
Any plans to add an ABRP (ABetterRoutePlanner) integration like TeslaLogger has?

**Comment by DrMichael** [comment]
Not sure. What is the benefit?

**Reply by dthiery** [reply]
Only need to give your Tesla acct info to one (self-hosted) service. Also prevents multiple services from simultaneously hitting the Tesla API which I think can cause sleep issues.

The ABRP integration itself gives realtime car info for route planning (battery state of charge, efficiency numbers, etc).

**Reply by LightBringer81** [reply]
Yes, I had ABRP linked with my Tesla Account but I removed the link, as ABRP woke up my car every time I opened ABRP for planning a route. With the possibility to link ABRP with TeslaMate we could use "real time" data read from TeslaMate, so if the car sleeps it won't need to be woke up for the latest stats.

**Comment by DrMichael** [comment]
Ok, that is what you get, when you give login in to Tesla in ABRP?

**Reply by dthiery** [reply]
Yea, and you get the same featureset through the TeslaLogger integration as well, but instead of providing ABRP with your Tesla login you provide TeslaLogger with an ABRP API token and then TeslaLogger makes an API call to ABRP periodically with an update on the car's status.

**Reply by DrMichael** [reply]
Given that we have an MQTT integration, wouldn't it make more sense to transport the information from mosquito to ABRP? So every *new* message gets send to ABRP.

**Reply by dthiery** [reply]
I'm sure that could be made to work. The easy no-code integration with ABRP is pretty slick from a user experience perspective given that ABRP is the go-to for route planning in the EV world.

TeslaLogger has a ScanMyTesla integration as well which is pretty cool. This is a different feature request, I know, but just thought I'd mention it here. Those two integrations (plus the fleet stats which I know you've mentioned you don't plan to incorporate in the past) are the big reasons to use TeslaLogger over TeslaMate right now IMO.

**Comment by dthiery** [comment]
If you decide you are interested in implementing this, I just took a peek at the relevant API call, and all that would needed is to POST the following on state change to https://api.iternio.com/1/tlm/send?token=$token&tlm=$json

{
"utc": utc,
"soc": soc
"speed": speed_kmh
"is_charging": is_charging ? 1 : 0
"power": power
"lat": lat
"lon": lon
}

The relevant TeslaLogger code can be found [here](https://github.com/bassmaster187/TeslaLogger/blob/61df9397ab21bf4abebfd25f0f256b2a4e2ede94/TeslaLogger/WebHelper.cs#L3626)

**Comment by adriankumpf** [comment]
If someone wants to look into it, I would accept a PR that contributes this feature.

**Comment by enigma2k** [comment]
it would be great if someone could make out of the teslalogger code a PR for Teslamate :)

**Comment by Powerbook001** [comment]
Hallo,
did somebody added this feature?
Thank you.

**Comment by letienne** [comment]
Did this. It's quick and ugly but it works.
Feel free to adapt to your needs.

Source: https://github.com/letienne/teslamate-abrp
Image: https://hub.docker.com/repository/docker/loaketienne/teslamate-abrp

Example from my compose file:

  teslamate-abrp:
    image: loaketienne/teslamate-abrp
    container_name: teslamate-abrp
    network_mode: bridge
    environment:
      - "MQTT_SERVER=1.2.3.4"
      - "API_KEY=yours"
      - "USER_TOKEN=yours"

**Reply by pratik5705** [reply]
What is the user token?

**Comment by Powerbook001** [comment]
Great,
you can tell me if it works also on NAS? i do not use Raspberry.

Thank yoy

**Reply by letienne** [reply]
I built it for my NAS (arch x86_64). 
Feel free to build it yourself if you need another arch. My dockerfile should work on ARM and others as well.

**Comment by letienne** [comment]
@adriankumpf Would you like me to polish this (add error management, documentation, all those nice things) and submit it as a PR?
I'm asking given that it uses completely different languages and solutions... I'm not sure that it is how you want to have it :)

**Reply by adriankumpf** [reply]
I simply don't have enough time at the moment. So if your PR needs little support and could be merged without big changes, it's more likely to be merged soon.
We'd then also need a little UI, where you can enter the API key and turn the integration on or off. The alternative would perhaps be to polish your existing project and offer it as an extension to TeslaMate that people could run in a separate container.

**Comment by OneThatWalks** [comment]
I am trying to write an ABRP integration in C# using the MQTT integration and I don't know why I am not seeing any messages.  I am using MQTT explorer and this is what I get when connecting to the docker host.  I am using the docker-compose.yml from the docs, am I missing something?

![image](https://user-images.githubusercontent.com/644628/125079030-2aed3300-e091-11eb-8224-5be43d809457.png)

**Comment by rawdigits** [comment]
I was just lamenting the fact that abrp was keeping my car awake again, when I ran across this timely update. It works great after... erm... borrowing an API key from another project on github. Given how well Teslamate works as a central source of all car data (alongside mqtt for access), this feels like a good thing to iterate on.


@letienne one small usability thing you may want to change in the script is allowing an env var to choose which car you're watching in MQTT. I sold a tesla, so `/car/1/*` doesn't have any data. It was easy to manually edit the script, but just a small idea.

Also if you want a PR, i'm happy to help a bit.

**Comment by Str1atum** [comment]
any news on that, is this in the official release yet?

**Comment by pleasanthill17** [comment]
Is there an option for mqtt username and password ?

**Comment by fetzu** [comment]
Hi everyone,

Seeing as I'd like to have this also (pulling car data for ABRP from TeslaMate) and that there has been some interest and groundwork laid on (on this discussion and on #2072), I'm willing to at least integrate this properly.

Right now, this would entail:

- [x] Refactoring some of the existing code (Dockerfile, python…)
- [x] Writing the documentation (and possibly add it to TeslaMate's)
- [x] Coordinating an “official” API key for TeslaMate from ABRP/Iternio (as they have asked according to #2072)
- [ ] (Depending on @adriankumpf answer to the next point) starting a PR

Now the greater question for you @adriankumpf is how you would want this integrated: 
1. Do you prefer to have something entirely separated (another Git, Docker image, etc. with its own documentation) that just leverages TeslaMate's MQTT server (and maybe a mention of it in the doc)?
2. Something “fully integrated” into TeslaMate (i.e.: have the documentation integrated as part of TeslaMate's, have a sub-folder in the repo with its own (Docker)files and use that to generate a “teslamate/teslamate-ABRP” docker image, add the needed service to the docker-compose in the doc, etc...) ?
3. Something in between (please let me know how you'd envision that)?

**Reply by fetzu** [reply]
@adriankumpf: I just saw your previous response (https://github.com/adriankumpf/teslamate/discussions/1592#discussioncomment-978866) and would try to find a way so that the user token (generated by ABRP) can be somehow set from TeslaMate's "Settings" page in this case. I don't have a clear solution for this right now, but I'll look into it. The API key itself should/will be the same for all TeslaMate users (pending on Iternio's response).

**Reply by dthiery** [reply]
Speaking purely as an end user with no affiliation to TeslaMate or @adriankumpf, I'd love to see something fully integrated with TeslaMate. Given the relatively minimal amount of code and overhead required for an integration like this, it seems silly to have to spin up a separate docker container with its own configuration just to pass a few data points over a REST api to ABRP. I suspect it would get a lot more use as a configuration option within TeslaMate itself (similar to how TeslaLogger handles their ABRP integration).

**Reply by Str1atum** [reply]
> Speaking purely as an end user with no affiliation to TeslaMate or @adriankumpf, I'd love to see something fully integrated with TeslaMate. Given the relatively minimal amount of code and overhead required for an integration like this, it seems silly to have to spin up a separate docker container with its own configuration just to pass a few data points over a REST api to ABRP. I suspect it would get a lot more use as a configuration option within TeslaMate itself (similar to how TeslaLogger handles their ABRP integration).

I would also vote for full integration in TeslaMate given the minimal overhead and usefulness.

**Reply by fetzu** [reply]
**Just so that we have a trace of this somewhere: I have received "TeslaMate's" API KEY for ABRP from Bo: d49234a5-2553-454e-8321-3fe30b49aa64.**

I have added the key to the Python-based implementation, and we are supposed to also use it for other ones.

**Reply by adriankumpf** [reply]
> Speaking purely as an end user with no affiliation to TeslaMate or @adriankumpf, I'd love to see something fully integrated with TeslaMate. Given the relatively minimal amount of code and overhead required for an integration like this, it seems silly to have to spin up a separate docker container with its own configuration just to pass a few data points over a REST api to ABRP. I suspect it would get a lot more use as a configuration option within TeslaMate itself (similar to how TeslaLogger handles their ABRP integration).

It may seem like little amount of code that would have to added initially. The problem with adding more and more functionality is that the maintenance burden keeps increasing and ultimately any future work (like bug fixes, API changes etc) – and also support requests (!) – fall to the maintainers of a project, and not to the initial contributor of a feature. 

Therefore, unsurprisingly, I prefer the first of your proposed options @fetzu :) Let's keep the projects separate. We can also promote it by adding to this page: https://docs.teslamate.org/docs/projects

**Comment by fetzu** [comment]
I have done a pass on @letienne code (via forks by @0ff, @pyjamasam and @pleasanthill17) and added a CLI and a Docker image that builds with the necessary instructions. I had a first contact with Bo from Iternio/ABRP, and he is ok with providing an API key for TeslaMate; I'll update the code and the doc as soon as I receive it. Code and instructions are available [here](https://github.com/fetzu/teslamate-abrp).

It is not the most elegant solution and the doc could probably be better, but it works. I'll have a look at integrating this directly in TeslaMate (in Elixir) without the need for an additional container; but I am not at all familiar with Elixir, so nobody should be holding their breath at this point ;).

**Reply by pleasanthill17** [reply]
Thanks for working on this !

**Reply by fetzu** [reply]
For the interested: the API key has now been added and the container should now run properly.

**Reply by letienne** [reply]
Thanks for keeping my code alive and taking care of it ! 🥇

**Reply by Str1atum** [reply]
Container install went smoothly without error messages but the car doesn't show up on ABRP after 15 min...
Portainer log shows

standard_init_linux.go:228: exec user process caused: exec format error

**Reply by fetzu** [reply]
> 
> 
> Container install went smoothly without error messages but the car doesn't show up on ABRP after 15 min... Portainer log shows
> 
> standard_init_linux.go:228: exec user process caused: exec format error

Are you by any chance trying to run the container on a non amd64 machine? What does "uname -a" (in the shell of the machine you are trying to run the container on) return?

**Reply by Str1atum** [reply]
Raspberry Pi 4 with Ubuntu-LTS:
Linux ubuntu 5.4.0-1045-raspi #49-Ubuntu SMP PREEMPT Wed Sep 29 17:49:16 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

**Reply by Str1atum** [reply]
Manual Python script startup bringt up the following:

```
./teslamate_mqtt2abrp.py: line 23: 
TeslaMate MQTT to ABRP

  Usage: 
    teslamate_mqtt2abrp.py [-hlap] [USER_TOKEN] [CAR_NUMBER] [CAR_MODEL] [MQTT_SERVER] [MQTT_USERNAME] [MQTT_PASSWORD]
  
  Arguments:
    USER_TOKEN          User token generated by ABRP.
    CAR_NUMBER          Car number from TeslaMate (usually 1).
    CAR_MODEL           Car model (from https://api.iternio.com/1/tlm/get_carmodels_list, e.g. s100d for a 2012-2018 S100D).
    MQTT_SERVER         MQTT server address (e.g. 192.168.1.1).
    MQTT_USERNAME       MQTT username (e.g. teslamate) - use with -l or -a.
    MQTT_PASSWORD       MQTT password (e.g. etamalset) - use with -a.

  Options:
    -h                  Show this screen.
    -l                  Use username to connect to MQTT server.
    -a                  Use authentification (user and password) to connect to MQTT server.

  Note:
    All arguments can also be passed as corresponding OS environment variables.
: File name too long
./teslamate_mqtt2abrp.py: line 26: import: command not found
./teslamate_mqtt2abrp.py: line 27: import: command not found
./teslamate_mqtt2abrp.py: line 28: import: command not found
./teslamate_mqtt2abrp.py: line 29: import: command not found
./teslamate_mqtt2abrp.py: line 30: import: command not found
from: can't read /var/mail/time
./teslamate_mqtt2abrp.py: line 32: import: command not found
./teslamate_mqtt2abrp.py: line 33: import: command not found
from: can't read /var/mail/docopt
./teslamate_mqtt2abrp.py: line 38: syntax error near unexpected token `('
./teslamate_mqtt2abrp.py: line 38: `    arguments = docopt(__doc__)' 
```

**Reply by iustin** [reply]
This simply looks like the script misses a hashbang and is considered a shell script? Just call python explicitly on it.

**Reply by Str1atum** [reply]
ok thanks, last Unix contact is quite some years ago.
Got the Python script running and working fine with ABRP.
Docker setup / container however still crashes.

**Reply by fetzu** [reply]
@Str1atum I’m glad you could get the python version working. The reason the containerized version is crashing is because the build targets the amd64 architecture. I tried building it for arm yesterday but it did not work. I’ll have another look and keep you updated here.

**Reply by fetzu** [reply]
@Str1atum: I created a build targeting arm64 that _should_ work. Can you try using "fetzu/teslamate-abrp-test" as an image and give me a quick feedback if it worked? If that's the case, I'll add arm64 to the “regular” build.

**Reply by Str1atum** [reply]
Access to that repository denied.

**Reply by fetzu** [reply]
Making it public would _probably_ help... sorry about that! It should be working now.

**Reply by Str1atum** [reply]
Image builds now, however a new error now:

Parameters same as direct Python script – server is 127.0.0.1



Traceback (most recent call last):

  File "/usr/src/teslamate-abrp/./teslamate_mqtt2abrp.py", line 105, in <module>
client.connect(MQTTSERVER)
File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 914, in connect
return self.reconnect()
File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 1044, in reconnect
sock = self._create_socket_connection()
File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 3685, in _create_socket_connection
return socket.create_connection(addr, timeout=self._connect_timeout, source_address=source)
File "/usr/local/lib/python3.10/socket.py", line 845, in create_connection
raise err

File "/usr/local/lib/python3.10/socket.py", line 833, in create_connection

sock.connect(sa)

ConnectionRefusedError: [Errno 111] Connection refused

**Reply by fetzu** [reply]
It looks like your MQTT server is refusing the connection.

1. Do you by any chance still have the python code running (i.e., the “non docker” one)?
2. Are you using some kind of special config for your MQTT server?
3. Try connecting to your MQTT server using MQTT explorer (https://mqtt-explorer.com/) or some other software to see if you manage to establish a connection and get some data back.

**EDIT:**
I didn't see your second sentence before, I guess 1 is cleared up.

Would you mind sharing your docker-compose? Don't forget to redact the secret bits ;).

**Reply by Str1atum** [reply]
Works.
Python script needs „127.0.0.1“ as server, docker container needs „mosquitto“ to make it work.
Thanks for the effort!

**Reply by fetzu** [reply]
That's what I suspected. I'm glad it works!

FYI I setup the default to build for both architectures, so feel free to revert back to fetzu/teslamate-abrp

**Reply by jAr4s** [reply]
I have used docker version. Set up proper car (Tesla Model 3 2022 60 LFP) and ABRP shows 100% SoC all the time. I can see in docker logs that this is what is being sent via MQTT - SoC 100%. Is there any setting that could affect that?

**Reply by fetzu** [reply]
Hey @jAr4s, could you please open an issue [here](https://github.com/fetzu/teslamate-abrp/issues) and I'll try to take a look at it this weekend.

If you could add a (censored) version of the MQTT message it would be great.

**Comment by Powerbook001** [comment]
> Not sure. What is the benefit?

Automatic route planning with real data from the car.

**Comment by pratik5705** [comment]
FWIW, I setup the ABRP integration, and to be honest, I turned it off. I
usually use ABRP to plan routes ahead of time and I usually do not need
current SOC. Others' use case might be different, but to me, it was not
that much valuable to have it.

**Reply by Powerbook001** [reply]
You can use it as a live navigation tools. Especially if you want to plan different chargers and route than the one Tesla offers. 
All in real time.
Not bad I would say!
